packages/htmlviewer/src/index.tsx

```diff
@@ -0,0 +1,260 @@
+    this._renderModel().then(() => (this._renderPending = false));
+  }
+
+  /**
+   * Render HTML in IFrame into this widget's node.
+   */
+  private async _renderModel(): Promise<void> {
+    let data = this.context.model.toString();
+    data = await this._setBase(data);
+
+    // Set the new iframe url.
+    const blob = new Blob([data], { type: 'text/html' });
+    const oldUrl = this._objectUrl;
+    this._objectUrl = URL.createObjectURL(blob);
+    this.content.url = this._objectUrl;
+
+    // Release reference to any previous object url.
+    if (oldUrl) {
+      try {
+        URL.revokeObjectURL(oldUrl);
+      } catch (error) {
+        /* no-op */
+      }
+    }
+    return;
+  }
+
+  /**
+   * Set a <base> element in the HTML string so that the iframe
+   * can correctly dereference relative links.
+   */
+  private async _setBase(data: string): Promise<string> {
+    const doc = this._parser.parseFromString(data, 'text/html');
+    let base: HTMLBaseElement;
+    base = doc.querySelector('base');
+    if (!base) {
+      base = doc.createElement('base');
+      doc.head.insertBefore(base, doc.head.firstChild);
+    }
+    const path = this.context.path;
+    const baseUrl = await this.context.urlResolver.getDownloadUrl(path);
+
+    // Set the base href, plus a fake name for the url of this
+    // document. The fake name doesn't really matter, as long
+    // as the document can dereference relative links to resources
+    // (e.g. CSS and scripts).
+    base.href = baseUrl;
+    base.target = '_self';
+    return doc.documentElement.innerHTML;
+  }
+
+  private _renderPending = false;
+  private _parser = new DOMParser();
+  private _monitor: ActivityMonitor<any, any> | null = null;
+  private _objectUrl: string = '';
+  private _trustedChanged = new Signal<this, boolean>(this);
+}
+
+/**
+ * A widget factory for HTMLViewers.
+ */
+export class HTMLViewerFactory extends ABCWidgetFactory<HTMLViewer> {
+  /**
+   * Create a new widget given a context.
+   */
+  protected createNewWidget(context: DocumentRegistry.Context): HTMLViewer {
+    return new HTMLViewer({ context });
+  }
+}
+
+/**
+ * A namespace for private data.
+ */
+namespace Private {
+  /**
+   * Sandbox exceptions for untrusted HTML.
+   */
+  export const untrusted: IFrame.SandboxExceptions[] = ['allow-same-origin'];
+
+  /**
+   * Sandbox exceptions for trusted HTML.
+   */
+  export const trusted: IFrame.SandboxExceptions[] = [
+    'allow-same-origin',
+    'allow-scripts'
+  ];
+
+  /**
+   * Namespace for TrustedButton.
+   */
+  export namespace TrustButtonComponent {
+    /**
+     * Interface for TrustedButton props.
+     */
+    export interface IProps {
+      htmlDocument: HTMLViewer;
+    }
+  }
+
+  /**
+   * React component for a trusted button.
+   *
+   * This wraps the ToolbarButtonComponent and watches for trust chagnes.
+   */
+  export function TrustButtonComponent(props: TrustButtonComponent.IProps) {
+    return (
+      <UseSignal
```

Comment:
👍

Reply:
I was skeptical, but it worked pretty well!